### ISSUE TITLE
Certificate SHA256 hashes are case-sensitive

### ISSUE THREAD

**Issue by p-harrison** [google/santa#1131]
I copy/pasted a certificate SHA256 hash from a Santa block popup message, which display in all caps like "EB 34 84 AD...", removed the spaces (so it became "EB3484AD...") and dropped it into my sync server to push down to devices.  The rule was added to the database but the application was still blocked.  I then used the SHA256 from "santactl fileinfo" which is all lowercase (so "eb3484ad...") and this worked successfully.

As these are hexadecimal I don't believe case should matter right?  As the Santa client displays both upper and lower case it should probably take a case-insensitive approach to rule matching as well.

**Comment by mlw** [comment]
Quick glance through the code confirms that the hashes generated for an executable file hash and certificate hash are lower case hex strings. If the rule contains upper case hex strings there will not be a match.

Due to how rule identifiers are stored in the database, it's difficult to migrate away from hex strings and instead do a binary comparison.

However, it does seem straight forward to ensure that when a given file or cert hash rule type is added, we could ensure we insert lowercase versions of the data.

**Comment by mlw** [comment]
> but what about also lowercasing anything read from the database before the comparison is done? This would ensure rules that people have accidentally added with uppercase characters (however unlikely) will now work as expected?

I'm not quite sure I follow the potential issue. The preference would be to store in the database in the final format desired during evaluation since that would happen on a hotter path and we'd want to minimize the work required there. If identifiers are translated when the rule is added to the database, this would ensure that no matter the format of identifier received, at evaluation time we would have the correct case.

> I also wonder if the Santa client GUI could display readily copy/pastable binary hash, cert hash and signingIDs so an end user can send a simple block of text in an email requesting an unblock, rather than having to use a command line tool which is beyond many users?

There is likely a fair request here in how we may better present information to users to be more easily consumable and potentially in formats that allow users to more easily request exceptions from administrators. However I'd like to ask to move that to a different issue if you could. I'm working on a patch for the case sensitivity issue described here and would not want to conflate them. Also it would be really helpful in the new issue to be very clear about what your needs are, what exact identifiers are missing or in an undesirable format in order to have a good discussion. Thanks!

**Comment by p-harrison** [comment]
> I'm not quite sure I follow the potential issue.

I was thinking of devices already in the field that might have rules added to their databases in uppercase - a change to the evaluation logic would get these rules functioning as expected.  But the next clean sync would tidy these up anyway if you make a change to how identifiers are written to the database, so all good.

**Comment by mlw** [comment]
I see what you mean now. The way I handled this was to bump the DB version and perform a migration of existing rule identifiers. This migration will occur when a host upgrades to the next Santa version and the process is restarted. A clean sync won't be required.
